### PR TITLE
Fix #1839: Fixed overlapping progressbar and hint bulb

### LINKED ISSUE #1839: Question player progress bar & hint bulb overlap unexpectedly
Question player progress bar & hint bulb overlap unexpectedly. See screenshot:

Ideally, we should make sure there's sufficient separation between the question progress bar & hint bulbs so that they don't overlap.

Note that this also seems to happen in the question player.

Final Solution as per below mentioned conversation.

**Scale Progress Bar to Fit** 

"Scale Progress Bar to Fit" will animate to the right once the hint button appears, which looks more natural flowing than if it were to move up.
As per the image above the `Progress Bar` size will change when the hint bulb appears and the `Progress Bar` size will change (going back to normal) once the hint bulb disappears. Note that this change should have animation in place so that it does not look abrupt change.

Reference: https://github.com/oppia/oppia-android/issues/1839#issuecomment-704698704

### PR DESCRIPTION
**Pull request by aishapervin**
## Explanation
Fixes #1839
i am adding a constraint from `progressbar` to` hints_and_solution_fragments_container` to solve this issue.

## Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

### CODE REVIEW

**Comment by aishapervin** [comment]
Screenshot_1611310463

**Review by FareesHussain — COMMENTED** [review]
Overall implementation looks good to me but I think this will be creating an issue with mobile-landscape as we are using padding start in the `ConstraintLayout` I suppose we need to add margin instead particularly for landscape for the following mentioned views and remove the paddingStart for the constraint layout to solve this @rt4914 what do you think ?
 - end_session_header_text_view
 - end_session_body_text_view
 - question_recycler_view

**Review by FareesHussain — CHANGES_REQUESTED** [review]
@aishapervin there seems to be some padding issue here you can append marginStart 24dp for the following views and remove to padding for the constraintLayout
 - end_session_header_text_view
- end_session_body_text_view
 - question_recycler_view

Adding Screenshot for reference
image

File: app/src/main/res/layout-land/question_player_fragment.xml [hunk 1]
```diff
@@ -121,30 +146,7 @@
...
-        android:visibility="@{viewModel.isHintOpenedAndUnRevealed() ? View.VISIBLE : View.GONE}" />
-
-      <ImageView
-        android:id="@+id/hint_bulb"
-        android:layout_width="wrap_content"
-        android:layout_height="wrap_content"
-        android:layout_margin="12dp"
-        android:src="@drawable/ic_hint_bulb_white_24dp" />
-    </FrameLayout>
+
```
Comment (prayutsu) [inline app/src/main/res/layout-land/question_player_fragment.xml]: This entire file needs reformatting.

File: app/src/main/res/layout-land/question_player_fragment.xml [hunk 2]
```diff
@@ -94,8 +94,33 @@
...
+          android:src="@drawable/ic_dot_yellow_24dp"
+          android:visibility="@{viewModel.isHintOpenedAndUnRevealed() ? View.VISIBLE : View.GONE}" />
+
+        <ImageView
+          android:id="@+id/hint_bulb"
+          android:layout_width="wrap_content"
+          android:layout_height="wrap_content"
+          android:layout_margin="12dp"
+          android:src="@drawable/ic_hint_bulb_white_24dp" />
+      </FrameLayout>
```
Comment (prayutsu) [inline app/src/main/res/layout-land/question_player_fragment.xml:124]: Any specific reason for shifting this entire framelayout above?

File: app/src/main/res/layout-sw600dp-land/question_player_fragment.xml [hunk 3]
```diff
@@ -95,7 +95,33 @@
...
+          android:src="@drawable/ic_dot_yellow_24dp"
+          android:visibility="@{viewModel.isHintOpenedAndUnRevealed() ? View.VISIBLE : View.GONE}" />
+
+        <ImageView
+          android:id="@+id/hint_bulb"
+          android:layout_width="wrap_content"
+          android:layout_height="wrap_content"
+          android:layout_margin="12dp"
+          android:src="@drawable/ic_hint_bulb_white_24dp" />
+      </FrameLayout>
```
Comment (prayutsu) [inline app/src/main/res/layout-sw600dp-land/question_player_fragment.xml:125]: Same applies here

File: app/src/main/res/layout-sw600dp-port/question_player_fragment.xml [hunk 4]
```diff
@@ -95,7 +95,34 @@
...
+          android:src="@drawable/ic_dot_yellow_24dp"
+          android:visibility="@{viewModel.isHintOpenedAndUnRevealed() ? View.VISIBLE : View.GONE}" />
+
+        <ImageView
+          android:id="@+id/hint_bulb"
+          android:layout_width="wrap_content"
+          android:layout_height="wrap_content"
+          android:layout_margin="12dp"
+          android:src="@drawable/ic_hint_bulb_white_24dp" />
+      </FrameLayout>
```
Comment (prayutsu) [inline app/src/main/res/layout-sw600dp-port/question_player_fragment.xml:131]: Same applies here

File: app/src/main/res/layout/question_player_fragment.xml [hunk 5]
```diff
@@ -93,8 +94,33 @@
         android:progressDrawable="@drawable/progress_bar"
         app:layout_constraintBottom_toTopOf="@id/question_progress_text"
         app:layout_constraintEnd_toEndOf="parent"
-        app:layout_constraintStart_toStartOf="parent" />
-
+        app:layout_constraintStart_toEndOf="@id/hints_and_solution_fragment_container" />
+      <FrameLayout
+        android:id="@+id/hints_and_solution_fragment_container"
+        android:layout_width="wrap_content"
```
Comment (prayutsu) [inline app/src/main/res/layout/question_player_fragment.xml:105]: Same applies here

**Review by prayutsu — CHANGES_REQUESTED** [review]
@aishapervin Left a few comments, PTAL, and all the suggested changes apply for all 4 files.

**Comment by prayutsu** [comment]
@rt4914 Not sure, but do we need a test case for the bug? Or we are not counting this as a bug?

File: app/src/main/res/layout-land/question_player_fragment.xml (hunk 2 above)
Comment (aishapervin) [reply app/src/main/res/layout-land/question_player_fragment.xml:124]: @prayutsu Because the frame layout is outside the constraint layout’s domain,we are not able to add constraints to progress bar for views outside of the constraint layout.So, just added this inside the constraint layout

File: app/src/main/res/layout-land/question_player_fragment.xml [hunk 6]
```diff
@@ -121,30 +146,7 @@
       android:layout_width="match_parent"
       android:layout_height="match_parent" />
```
Comment (prayutsu) [inline app/src/main/res/layout-land/question_player_fragment.xml:150]: In the current implementation, the `hints_and_solution_fragment_placeholder` frameLayout is of no use. So, you can delete this layout and after that, you would not require to shift `hints_and_solution_fragment_container` above.

File: app/src/main/res/layout-land/question_player_fragment.xml (hunk 2 above)
Comment (prayutsu) [reply app/src/main/res/layout-land/question_player_fragment.xml:124]: Please see [this](https://github.com/oppia/oppia-android/pull/2540/files#r562765531)

File: app/src/main/res/layout-land/question_player_fragment.xml (hunk 1 above)
Comment (prayutsu) [reply app/src/main/res/layout-land/question_player_fragment.xml]: @aishapervin You need to press `Ctrl + Shift + Alt + L` to auto-format this complete file at once so that we can get rid of empty lines and other formatting errors.
This applies for all the 4 files.

**Comment by prayutsu** [comment]
@aishapervin Please note that it is a good practice to add a short description of the changes in the PR description as mentioned in the first point of the checklist.

File: app/src/main/res/layout-land/question_player_fragment.xml (hunk 1 above)
Comment (aishapervin) [reply app/src/main/res/layout-land/question_player_fragment.xml]: done

File: app/src/main/res/layout-land/question_player_fragment.xml (hunk 6 above)
Comment (aishapervin) [reply app/src/main/res/layout-land/question_player_fragment.xml:150]: Even after removing the "hints_and_solution_fragment_placeholder" i would need to add the "hints_and_solution_fragment_container" inside the constraint layout as I'm adding a constraint from progressbar to  "hints_and_solution_fragment_container"

Comment (prayutsu) [reply app/src/main/res/layout-land/question_player_fragment.xml:150]: @aishapervin Definitely, you are right but we want to get rid of `hints_and_solution_fragment_placeholder`, so it needs to get deleted and yes you'll have to shift it up in the constraint layout.

**Comment by aishapervin** [comment]
> @aishapervin Please note that it is a good practice to add a short description of the changes in the PR description as mentioned in the first point of the checklist.

Thanks ,i have updated the pr description.

File: app/src/main/res/layout-land/question_player_fragment.xml (hunk 6 above)
Comment (aishapervin) [reply app/src/main/res/layout-land/question_player_fragment.xml:150]: done.

File: app/src/main/res/layout-land/question_player_fragment.xml (hunk 2 above)
Comment (aishapervin) [reply app/src/main/res/layout-land/question_player_fragment.xml:124]: followed up in below comments.

File: app/src/main/res/layout-sw600dp-land/question_player_fragment.xml (hunk 3 above)
Comment (aishapervin) [reply app/src/main/res/layout-sw600dp-land/question_player_fragment.xml:125]: followed up in below comments.

File: app/src/main/res/layout-sw600dp-port/question_player_fragment.xml (hunk 4 above)
Comment (aishapervin) [reply app/src/main/res/layout-sw600dp-port/question_player_fragment.xml:131]: followed up in below comments.

File: app/src/main/res/layout/question_player_fragment.xml (hunk 5 above)
Comment (aishapervin) [reply app/src/main/res/layout/question_player_fragment.xml:105]: followed up in below comments.

**Review by FareesHussain — APPROVED** [review]
Except landscape layout everything looks good to me 
@rt4914 can you please have a look at the padding and margin changes in the
`layout-land/question_player_fragment.xml`

**Comment by rt4914** [comment]
Will check this on Monday. Meanwhile you can started working on some other issue. Thanks @aishapervin

**Review by prayutsu — APPROVED** [review]
Other than what @FareesHussain mentioned, the PR looks good to me

**Review by rt4914 — CHANGES_REQUESTED** [review]
This Pr changes a lot of padding and margin. For which I will need check it with correct mocks. Will need some extra time on this.

File: app/src/main/res/layout-land/question_player_fragment.xml [hunk 7]
```diff
@@ -84,23 +84,50 @@
...
-        android:layout_marginEnd="30dp"
+        android:layout_marginEnd="54dp"
         android:layout_marginBottom="5dp"
         android:max="100"
         android:progress="@{viewModel.progressPercentage}"
         android:progressDrawable="@drawable/progress_bar"
         app:layout_constraintBottom_toTopOf="@id/question_progress_text"
         app:layout_constraintEnd_toEndOf="parent"
-        app:layout_constraintStart_toStartOf="parent" />
+        app:layout_constraintStart_toEndOf="@id/hints_and_solution_fragment_container" />
```
Comment (rt4914) [inline app/src/main/res/layout-land/question_player_fragment.xml]: For landscape this should not be constrained.
Updated this to `app:layout_constraintStart_toStartOf="parent" />`
and marginStart to `54dp`.

This way the progress bar will always look in center.

Also, because the size of the `hints_and_solution_fragment_container ` is not more than `54dp` it will never overlap in landscape mode.

File: app/src/main/res/layout-land/question_player_fragment.xml [hunk 8]
```diff
@@ -116,36 +143,6 @@
         tools:layout_constraintGuide_percent="0.5f" />
     </androidx.constraintlayout.widget.ConstraintLayout>
 
-    <FrameLayout
```
Comment (rt4914) [inline app/src/main/res/layout-land/question_player_fragment.xml]: Why was this removed? We need this.

File: app/src/main/res/layout-sw600dp-land/question_player_fragment.xml [hunk 9]
```diff
@@ -95,7 +95,34 @@
         android:progressDrawable="@drawable/progress_bar"
         app:layout_constraintBottom_toTopOf="@id/question_progress_text"
         app:layout_constraintEnd_toEndOf="parent"
-        app:layout_constraintStart_toStartOf="parent" />
+        app:layout_constraintStart_toEndOf="@id/hints_and_solution_fragment_container" />
```
Comment (rt4914) [inline app/src/main/res/layout-sw600dp-land/question_player_fragment.xml]: Update this to `app:layout_constraintStart_toStartOf="parent" />`
 because marginStart=96 and marginEnd=96dp makes sure that the progressbar never overlaps with the `hints_and_solution_fragment_container `

So technically we don't need any change in this file except that of `android:layout_width="0dp"` in ProgressBar

File: app/src/main/res/layout-sw600dp-port/question_player_fragment.xml [hunk 10]
```diff
@@ -95,7 +95,34 @@
         android:progressDrawable="@drawable/progress_bar"
         app:layout_constraintBottom_toTopOf="@id/question_progress_text"
         app:layout_constraintEnd_toEndOf="parent"
-        app:layout_constraintStart_toStartOf="parent" />
+        app:layout_constraintStart_toEndOf="@id/hints_and_solution_fragment_container" />
```
Comment (rt4914) [inline app/src/main/res/layout-sw600dp-port/question_player_fragment.xml]: Update this to `app:layout_constraintStart_toStartOf="parent" />`
 because marginStart=64 and marginEnd=64dp makes sure that the progressbar never overlaps with the `hints_and_solution_fragment_container `

So technically we don't need any change in this file except that of `android:layout_width="0dp"` in ProgressBar

File: app/src/main/res/layout-sw600dp-port/question_player_fragment.xml [hunk 11]
```diff
@@ -120,36 +147,6 @@
         tools:layout_constraintGuide_percent="0.5f" />
     </androidx.constraintlayout.widget.ConstraintLayout>
 
-    <FrameLayout
```
Comment (rt4914) [inline app/src/main/res/layout-sw600dp-port/question_player_fragment.xml:123]: Add this back.

File: app/src/main/res/layout/question_player_fragment.xml [hunk 12]
```diff
@@ -118,36 +145,6 @@
         tools:layout_constraintGuide_percent="0.5f" />
     </androidx.constraintlayout.widget.ConstraintLayout>
 
-    <FrameLayout
```
Comment (rt4914) [inline app/src/main/res/layout/question_player_fragment.xml:121]: Add this back

**Review by rt4914 — CHANGES_REQUESTED** [review]
@aishapervin Thanks for the PR.
Major changes:
1. In `layout/question_player_fragment` your logic is correct which constraints both the UIs.
2. In all other files just set `layout_width` to `0dp` for `ProgressBar` and all other changes should be reverted.

File: app/src/main/res/layout-sw600dp-port/question_player_fragment.xml (hunk 11 above)
Comment (prayutsu) [reply app/src/main/res/layout-sw600dp-port/question_player_fragment.xml:123]: @aishapervin Apologies for the confusion here, I thought it isn't needed here. Thanks, @rt4914

File: app/src/main/res/layout-land/question_player_fragment.xml (hunk 8 above)
Comment (prayutsu) [reply app/src/main/res/layout-land/question_player_fragment.xml]: @aishapervin Apologies for the confusion here, I thought it isn't needed here. Thanks, @rt4914

**Comment by Luffy18346** [comment]
@aishapervin If you have done mentioned changes by reviewer then please write "Done" or "Changes are done" in reply. **Remember, Do not resolve the conversations,** they will be resolved by the reviewer once he/she verifies and okays the changes done by you.

File: app/src/main/res/layout-land/question_player_fragment.xml (hunk 7 above)
Comment (aishapervin) [reply app/src/main/res/layout-land/question_player_fragment.xml]: done

File: app/src/main/res/layout-land/question_player_fragment.xml (hunk 8 above)
Comment (aishapervin) [reply app/src/main/res/layout-land/question_player_fragment.xml]: done

File: app/src/main/res/layout-sw600dp-land/question_player_fragment.xml (hunk 9 above)
Comment (aishapervin) [reply app/src/main/res/layout-sw600dp-land/question_player_fragment.xml]: okk.done

File: app/src/main/res/layout-sw600dp-port/question_player_fragment.xml (hunk 10 above)
Comment (aishapervin) [reply app/src/main/res/layout-sw600dp-port/question_player_fragment.xml]: done

File: app/src/main/res/layout-sw600dp-port/question_player_fragment.xml (hunk 11 above)
Comment (aishapervin) [reply app/src/main/res/layout-sw600dp-port/question_player_fragment.xml:123]: changes are done

File: app/src/main/res/layout/question_player_fragment.xml (hunk 12 above)
Comment (aishapervin) [reply app/src/main/res/layout/question_player_fragment.xml:121]: done

**Comment by Luffy18346** [comment]
@rt4914 PTAL

File: app/src/main/res/layout-land/question_player_fragment.xml (hunk 7 above)
Comment (rt4914) [reply app/src/main/res/layout-land/question_player_fragment.xml]: This is still not correct. As per above comment, change the constraint to
`app:layout_constraintStart_toStartOf="parent"`
and remove 
`app:layout_constraintStart_toEndOf="@id/hints_and_solution_fragment_container"`

File: app/src/main/res/layout-sw600dp-land/question_player_fragment.xml (hunk 9 above)
Comment (rt4914) [reply app/src/main/res/layout-sw600dp-land/question_player_fragment.xml]: This is still not correct.
Remove `app:layout_constraintStart_toEndOf="@id/hints_and_solution_fragment_container"`
and add `app:layout_constraintStart_toStartOf="parent"` and make ProgressBar width `android:layout_width="0dp"`

File: app/src/main/res/layout-sw600dp-port/question_player_fragment.xml (hunk 10 above)
Comment (rt4914) [reply app/src/main/res/layout-sw600dp-port/question_player_fragment.xml]: Ditto as above.

**Review by rt4914 — CHANGES_REQUESTED** [review]
@aishapervin I have replied to the comments. 
In general wherever I had written `Update` then meant that you had to update that specific like with the suggestion that I gave. Instead you have kept the earlier line and added the suggestion that I gave which is incorrect.

Thanks.

**Comment by aishapervin** [comment]
> @aishapervin I have replied to the comments.
> In general wherever I had written `Update` then meant that you had to update that specific like with the suggestion that I gave. Instead you have kept the earlier line and added the suggestion that I gave which is incorrect.
> 
> Thanks.

@rt4914 changes are updated.

File: app/src/main/res/layout-land/question_player_fragment.xml [hunk 13]
```diff
@@ -96,11 +96,38 @@
         app:layout_constraintEnd_toEndOf="parent"
         app:layout_constraintStart_toStartOf="parent" />
 
+      <FrameLayout
```
Comment (rt4914) [inline app/src/main/res/layout-land/question_player_fragment.xml:99]: These changes are not correct as per this comment: https://github.com/oppia/oppia-android/pull/2540#pullrequestreview-578738098

File: app/src/main/res/layout-sw600dp-land/question_player_fragment.xml [hunk 14]
```diff
@@ -120,36 +147,6 @@
         tools:layout_constraintGuide_percent="0.5f" />
     </androidx.constraintlayout.widget.ConstraintLayout>
 
-    <FrameLayout
```
Comment (rt4914) [inline app/src/main/res/layout-sw600dp-land/question_player_fragment.xml:123]: These changes are not correct as per this comment: https://github.com/oppia/oppia-android/pull/2540#pullrequestreview-578738098

Also you have removed this `hints_and_solution_fragment_placeholder ` which means that `Hints&Solution` won't work in tablet-landscape

File: app/src/main/res/layout-sw600dp-port/question_player_fragment.xml [hunk 15]
```diff
@@ -97,6 +97,38 @@
         app:layout_constraintEnd_toEndOf="parent"
         app:layout_constraintStart_toStartOf="parent" />
 
+      <FrameLayout
```
Comment (rt4914) [inline app/src/main/res/layout-sw600dp-port/question_player_fragment.xml:100]: These changes are not correct as per this comment: https://github.com/oppia/oppia-android/pull/2540#pullrequestreview-578738098

**Review by rt4914 — CHANGES_REQUESTED** [review]
@aishapervin The changes are still not correct.

Suggest closing this PR and creating a new PR focusing only on this comment and that would be easier:
https://github.com/oppia/oppia-android/pull/2540#pullrequestreview-578738098

**Comment by aishapervin** [comment]
created #2688 to fix this.
